File: src/lib/features/volume.ts
// Volume score calculation

import { calcVolumeMA } from "./calculator";

export interface VolumeResult {
  score: number;
  detail: {
    volume_current: number;
    volume_ma20: number;
    volume_ratio: number;
    days_used: number;
  };
}

/**
 * Calculate volume score based on current volume vs MA20
 */
export function calculateVolumeScore(volumes: number[]): VolumeResult {
  if (volumes.length === 0) {
    return {
      score: 50,
      detail: {
        volume_current: 0,
        volume_ma20: 0,
        volume_ratio: 1,
        days_used: 0,
      },
    };
  }

  const current = volumes[volumes.length - 1];
  const ma20 = calcVolumeMA(volumes, 20);
  const ratio = ma20 > 0 ? current / ma20 : 1;

  const score = scoreVolumeRatio(ratio);

  return {
    score,
    detail: {
      volume_current: Number(current.toFixed(2)),
      volume_ma20: Number(ma20.toFixed(2)),
      volume_ratio: Number(ratio.toFixed(3)),
      days_used: Math.min(20, volumes.length),
    },
  };
}

function scoreVolumeRatio(ratio: number): number {
  if (ratio > 3.0) return 95;
  if (ratio > 2.0) return 85;
  if (ratio > 1.5) return 75;
  if (ratio > 1.0) return 60;
  if (ratio > 0.7) return 45;
  if (ratio > 0.5) return 30;
  return 15;
}

File: backend/features/volume.py
"""
Volume score calculation
"""
import pandas as pd
from typing import Dict, Any
from backend.features.calculator import calc_volume_ma


def calculate_volume_score(df: pd.DataFrame) -> Dict[str, Any]:
    """
    Calculate volume score based on current volume vs MA20
    
    Input: DataFrame với cột 'volume', ít nhất 1 row
    """
    if len(df) == 0 or 'volume' not in df.columns:
        return {
            "score": 50.0,
            "detail": {
                "volume_current": 0,
                "volume_ma20": 0,
                "volume_ratio": 1.0,
                "days_used": 0,
            }
        }

    volumes = df['volume'].astype(float)
    current = float(volumes.iloc[-1])
    ma20 = calc_volume_ma(volumes, period=20)
    ratio = current / ma20 if ma20 > 0 else 1.0

    score = _score_volume_ratio(ratio)

    return {
        "score": float(score),
        "detail": {
            "volume_current": round(current, 2),
            "volume_ma20": round(ma20, 2),
            "volume_ratio": round(ratio, 3),
            "days_used": min(20, len(volumes)),
        }
    }


def _score_volume_ratio(ratio: float) -> float:
    if ratio > 3.0:   return 95.0
    if ratio > 2.0:   return 85.0
    if ratio > 1.5:   return 75.0
    if ratio > 1.0:   return 60.0
    if ratio > 0.7:   return 45.0
    if ratio > 0.5:   return 30.0
    return 15.0
